### PR TITLE
Generate API Docs

### LINKED ISSUE #239: Generate API object reference documentation
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

We need to document our configuration objects in a better and higher automated way. W

**What you expected to happen**:

We should use the same tooling as Kubernetes upstream.

Kube builder provides a good doc for it: 

http://book.kubebuilder.io/beyond_basics/generating_documentation.html

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration**:
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by JoshVanL**
What this PR does / why we need it:
Adds Scripts to generate API documentation for the site and makes sure it's up to date in CI.

Special notes for your reviewer:
Many API type definitions violated the generators validations so have been changed to follow convention.

fixes #239 

```release-note
Generate API documentation in site
```
/assign @simonswine

### CODE REVIEW

**Comment by simonswine** [comment]
Also the generated code looks a bit different for me can you address that, it should fix that:  `npm notice created a lockfile as package-lock.json. You should commit this file.`

**Comment by JoshVanL** [comment]
/unassign
/assign @simonswine 
@simonswine This should fix these issues, LMK if you still have problems

File: Makefile [hunk 1]
```diff
@@ -20,18 +20,26 @@ GOPATH ?= /tmp/go
...
 ifeq ($(UNAME_S),Darwin)
+	TAR_EXT := tar.gz
 	SHASUM := shasum -a 256 -c
 	DEP_URL := https://github.com/golang/dep/releases/download/v0.5.0/dep-darwin-amd64
 	DEP_HASH := 1a7bdb0d6c31ecba8b3fd213a1170adf707657123e89dff234871af9e0498be2
 	GORELEASER_URL := https://github.com/goreleaser/goreleaser/releases/download/v0.77.0/goreleaser_Darwin_x86_64.tar.gz
 	GORELEASER_HASH := bc6cdf2dfe506f2cce5abceb30da009bfd5bcdb3e52608c536e6c2ceea1f24fe
+    NODE_NAME := node-v8.12.0-darwin-x64
+    NODE_URL :=	https://nodejs.org/dist/v8.12.0/${NODE_NAME}.$(TAR_EXT)
+    NODE_HASH := ca131b84dfcf2b6f653a6521d31f7a108ad7d83f4d7e781945b2eca8172064aa
```
Comment (simonswine) [inline Makefile]: Can we use the same whitespaces here?

File: Makefile [hunk 2]
```diff
@@ -20,18 +20,26 @@ GOPATH ?= /tmp/go
...
 ifeq ($(UNAME_S),Linux)
+	TAR_EXT := tar.xz
 	SHASUM := sha256sum -c
 	DEP_URL := https://github.com/golang/dep/releases/download/v0.5.0/dep-linux-amd64
 	DEP_HASH := 287b08291e14f1fae8ba44374b26a2b12eb941af3497ed0ca649253e21ba2f83
 	GORELEASER_URL := https://github.com/goreleaser/goreleaser/releases/download/v0.77.0/goreleaser_Linux_x86_64.tar.gz
 	GORELEASER_HASH := aae3c5bb76b282e29940f2654b48b13e51f664368c7589d0e86b391b7ef51cc8
+    NODE_NAME := node-v8.12.0-linux-x64
+    NODE_URL := https://nodejs.org/dist/v8.12.0/${NODE_NAME}.$(TAR_EXT)
+    NODE_HASH := 29a20479cd1e3a03396a4e74a1784ccdd1cf2f96928b56f6ffa4c8dae40c88f2
```
Comment (simonswine) [inline Makefile]: Can we use the same whitespaces here?

File: docs/conf.py [hunk 3]
```diff
@@ -42,14 +42,15 @@
 # The suffix(es) of source filenames.
 # You can specify multiple suffix as a list of string:
 #
-source_suffix = ['.rst', '.md']
+#source_suffix = ['.rst', '.md']
+source_suffix = ['.rst']
```
Comment (simonswine) [inline docs/conf.py]: :heart_eyes:

File: docs/conf.py [hunk 4]
```diff
@@ -42,14 +42,15 @@
...
 
 # The master toctree document.
 master_doc = 'index'
 
 # General information about the project.
 project = u'Tarmak'
-copyright = u'2017, Jetstack'
+copyright = u'2018, Jetstack'
```
Comment (simonswine) [inline docs/conf.py]: Maybe we solve it for all time:

```
import time
copyright = u'%s, Jetstack' % time.strftime('%Y')
```

File: Makefile [hunk 5]
```diff
@@ -142,24 +160,38 @@ endif
 
 $(BINDIR)/goreleaser:
 	curl -sL -o $@.tar.gz $(GORELEASER_URL)
-	echo "$(GORELEASER_HASH)  $@.tar.gz" | $(SHASUM)
+	echo "$(GORELEASER_HASH) $@.tar.gz" | $(SHASUM)
```
Comment (MattiasGees) [inline Makefile]: Add the extra space again. This is needed to make it work on MacOS.

File: Makefile [hunk 6]
```diff
@@ -124,9 +132,19 @@ $(BINDIR)/informer-gen:
 
 $(BINDIR)/dep:
 	curl -sL -o $@ $(DEP_URL)
-	echo "$(DEP_HASH)  $@" | $(SHASUM)
+	echo "$(DEP_HASH) $@" | $(SHASUM)
```
Comment (MattiasGees) [inline Makefile]: Add the extra space again. This is needed to make it work on MacOS.

File: Makefile [hunk 7]
```diff
@@ -124,9 +132,19 @@ $(BINDIR)/informer-gen:
...
 	chmod +x $@
 
+$(BINDIR)/node:
+	curl -sL -o $(BINDIR)/$(NODE_NAME).$(TAR_EXT) $(NODE_URL)
+	echo "$(NODE_HASH) $(BINDIR)/$(NODE_NAME).$(TAR_EXT)" | $(SHASUM)
```
Comment (MattiasGees) [inline Makefile]: Add extra space between `$(NODE_HASH)` and `$(BINDIR)/$(NODE_NAME).$(TAR_EXT)`. This is needed to make it work on MacOS.

**Comment by MattiasGees** [comment]
One difference with the committed output for me is:
```
	typechange: docs/generated/reference/output/node_modules/.bin/marked
	typechange: docs/generated/reference/output/node_modules/.bin/mime
	typechange: docs/generated/reference/output/node_modules/.bin/static
```

These became files instead of symlinks.

**Comment by simonswine** [comment]
/unassign @MattiasGees 

Thanks @JoshVanL, I will rebase this

**Comment by JoshVanL** [comment]
@simonswine Sorry didn't see your message, went ahead and did it

**Comment by simonswine** [comment]
Thank you @JoshVanL this is a great improvement!

/approve
/lgtm

**Pull request merged**